leshan-core/src/main/java/org/eclipse/leshan/core/attributes/AttributeSet.java

```diff
@@ -0,0 +1,144 @@
+            if (!attr.canBeAssignedTo(assignationLevel)) {
+                throw new IllegalArgumentException(String.format("Attribute '%s' cannot be assigned to level %s",
+                        attr.getCoRELinkParam(), assignationLevel.name()));
+            }
+        }
+        Attribute pmin = attributeMap.get(AttributeName.MINIMUM_PERIOD);
+        Attribute pmax = attributeMap.get(AttributeName.MAXIMUM_PERIOD);
+        if ((pmin != null) && (pmax != null) && (Long) pmin.getValue() > (Long) pmax.getValue()) {
+            throw new IllegalArgumentException(String.format("Cannot write attributes where '%s' > '%s'",
+                    pmin.getCoRELinkParam(), pmax.getCoRELinkParam()));
+        }
+    }
+
+
+    /**
+     * Returns a new AttributeSet, containing only the attributes that have a matching
+     * Attachment level. 
+     * @param attachment the Attachment level to filter by
+     * @return a new {@link AttributeSet} containing the filtered attributes
+     */
+    public AttributeSet filter(Attachment attachment) {
+        List<Attribute> attrs = new ArrayList<>();
+        for (Attribute attr : getAttributes()) {
+            if (attr.getAttachment() == attachment) {
+                attrs.add(attr);
+            }
+        }
+        return new AttributeSet(attrs.toArray(new Attribute[attrs.size()]));
+    }
+    
+    /**
+     * Creates a new AttributeSet by merging another AttributeSet onto this instance.
+     * @param attributes the AttributeSet that should be merged onto this instance. Attributes in this
+     * set will overwrite existing attribute values, if present. If this is null, the
+     * new attribute set will effectively be a clone of the existing one
+     * @return the merged AttributeSet 
+     */
+    public AttributeSet merge(AttributeSet attributes) {
+        Map<AttributeName, Attribute> merged = new LinkedHashMap<>();
+        for (Attribute attr : getAttributes()) {
+            merged.put(attr.getName(), attr);
+        }
+        if (attributes != null) {
+            for (Attribute attr : attributes.getAttributes()) {
+                merged.put(attr.getName(), attr);
+            }
+        }
+        return new AttributeSet(merged.values().toArray(new Attribute[merged.size()]));        
+    }
+
+    /**
+     * Returns the attributes as a map with the CoRELinkParam as key and the attribute value as map value.
+     * @return the attributes map
+     */
+    public Map<String, Object> getMap() {
+        Map<String, Object> result = new LinkedHashMap<>();
+        for (Attribute attr : attributeMap.values()) {
+            result.put(attr.getCoRELinkParam(), attr.getValue());
+        }
+        return result;
+    }
+    
+    public Attribute[] getAttributes() {
+        return attributeMap.values().toArray(new Attribute[attributeMap.size()]);
+    }
+    
+    public String[] toQueryParams() {
+        List<String> queries = new LinkedList<>();
```

Comment:
You can directly create the array as you know the size.

Reply:
Yes, but then I would have to use a for (int x = 0... loop, which in my mind is harder to read